# Notebook E-tivity 1 CE4021

Student name: Jason Coleman

Student ID: 9539719

## Imports

In [1]:
#None

If you believe required imports are missing, please contact your moderator.

<hr style="border:2px solid gray"> </hr>

## Task - Option 1: Symbolic calculation of the derivative
Write 2 Python functions:

* The first one to *symbolically calculate the derivative of a polynomial with one variable*. The input should be a polynomial and the output should also be a polynomial.
* The second one to *evaluate (i.e. get the numerical value) of a polynomial for a given value of its variable*. The input should be a polynomial and a value (point at which to evaluate the polynomial). The output should be a scalar.
* Test your code with a few salient polynomials (minimum of 3) for which you have calculated the derivative of these polynomials manually.

You may add as many cells as you require to complete this task. Refer to the Rubric for E-tivity 1 to see how you will be assessed.

## Polynomial Evaluation

We can evaluate a polynomial in terms of x by simply evaluating each element of the set of tuples and summing the result

$$P(x) = a_n x^n + a_{n-1} x^{n-1} + \ldots + a_2 x^2 + a_1 x + a_0$$
So, for $p=2$ we would have
$$P(2) = a_n \cdot 2^n + a_{n-1} \cdot 2^{n-1} + \ldots + a_2 \cdot 2^2 + a_1 \cdot 2 + a_0$$

In [2]:
def evaluate_polynomial(elements: set[tuple[int, int]], value: int) -> int:
    """
    Evaluate a polynomial in x, represented by a set of coefficient-power tuples at x.
    :param elements: set of coefficients-power tuples
    :param value: value to evaluate the polynomial at
    :return: value of the polynomial at x
    """
    s = 0
    for coefficient, power in elements:
        s += coefficient * value ** power

    return s

## Differentiation
The python code executes the Power rule on each element of the set of tuples. 

$$\frac{d}{dx}(coefficient \cdot x^{power}) = coefficient \cdot power \cdot x^{(power-1)}$$

And returns the set of the derivatures that make up the output polynomial.

In [3]:
def differentiate(elements: set[tuple[int, int]]) -> set[tuple[int, int]]:
    """
    Differentiate a polynomial in x, with respect to x using the power rule.
    Where the polynomial is represented by a set of coefficients-power tuples.
    Currently, it uses a more loop-based approach for ease-of-reading.
    :param elements: coefficients-power tuples of the polynomial
    :return: coefficients of the differentiated polynomial
    """
    result = set()
    for coefficient, power in elements:
        if power != 0:
            result.add((coefficient * power, power - 1))
        else:
            result.add((0, 0))
    return result

## Utility methods: Pretty print
The following is a utility function to make the polynomials more human-readable.

In [4]:
def to_str(coefficients_: set[tuple[int, int]]) -> str:
    """
    A simple utility function. Converts a set of coefficient-power tuples
    to a string representation of a polynomial.  Print coefficients
    as a string (e.g. "5x^3 + 7x^2 + 3x + 6")
    :param coefficients_: a set of coefficient-power tuples
    :return: str representation of the polynomial
    """
    coefficients_str = ""

    # Sort coefficients by power in descending order - abs(power) in case of negative powers
    # This will make the final string easier to read - or closer to conventional notation
    coefficients_ = sorted(coefficients_, key=lambda x_: abs(x_[1]), reverse=True)

    if coefficients_ == [(0, 0)]:
        return "0"

    for coefficient, power in coefficients_:
        if coefficient == 1:
            sign = ''
        elif coefficient == -1:
            sign = '-'
        else:
            sign = f'{coefficient}'

        if power > 1 or power < 0:
            coefficients_str += f"{sign}x^{power} + "
        elif power == 1:
            coefficients_str += f"{sign}x + "
        else:  # power == 0
            if coefficient != 0:
                coefficients_str += f"{coefficient} + "

    # Remove the trailing "+ " at the end
    coefficients_str = coefficients_str.rstrip("+ ")

    # handle the case where coefficient is negative - for ease of reading
    return coefficients_str.replace("+ -", "- ")

Note that this is the most complex part of this assignment; where most things can go wrong.

## Examples of use
Here we show the use of the derivative and evaluation functions (as well as the utility to_str).

#### Example 1

In [5]:
polynomial = {(5, 3), (7, 2), (3, 1), (6, 0)}
print(f"f(x): {to_str(polynomial)}")

x = 2
f_x = evaluate_polynomial(polynomial, x)
print(f"f({x}): {f_x}")

f_prime = differentiate(polynomial)
print(f"f'(x): {to_str(f_prime)}")

f_prime_x = evaluate_polynomial(f_prime, x)
print(f"f'({x}): {f_prime_x}")

f(x): 5x^3 + 7x^2 + 3x + 6
f(2): 80
f'(x): 15x^2 + 14x + 3
f'(2): 91


#### Example 2

In [6]:
polynomial =  {(2, 1000)}

print(f"f(x): {to_str(polynomial)}")

x = 2
f_x = evaluate_polynomial(polynomial, x)
print(f"f({x}): {f_x}")

f_prime = differentiate(polynomial)
print(f"f'(x): {to_str(f_prime)}")

f_prime_x = evaluate_polynomial(f_prime, x)
print(f"f'({x}): {f_prime_x}")

f(x): 2x^1000
f(2): 21430172143725346418968500981200036211228096234110672148875007767407021022498722449863967576313917162551893458351062936503742905713846280871969155149397149607869135549648461970842149210124742283755908364306092949967163882534797535118331087892154125829142392955373084335320859663305248773674411336138752
f'(x): 2000x^999
f'(2): 10715086071862673209484250490600018105614048117055336074437503883703510511249361224931983788156958581275946729175531468251871452856923140435984577574698574803934567774824230985421074605062371141877954182153046474983581941267398767559165543946077062914571196477686542167660429831652624386837205668069376000


#### Example 3

In [7]:
polynomial =   {(3, 2), (-4, 1), (2, 0)}

print(f"f(x): {to_str(polynomial)}")

x = 2
f_x = evaluate_polynomial(polynomial, x)
print(f"f({x}): {f_x}")

f_prime = differentiate(polynomial)
print(f"f'(x): {to_str(f_prime)}")

f_prime_x = evaluate_polynomial(f_prime, x)
print(f"f'({x}): {f_prime_x}")

f(x): 3x^2 - 4x + 2
f(2): 6
f'(x): 6x - 4
f'(2): 8


#### Example 4

In [8]:
polynomial =  {(1, 5), (2, 3), (-1, 2)}

print(f"f(x): {to_str(polynomial)}")

x = 2
f_x = evaluate_polynomial(polynomial, x)
print(f"f({x}): {f_x}")

f_prime = differentiate(polynomial)
print(f"f'(x): {to_str(f_prime)}")

f_prime_x = evaluate_polynomial(f_prime, x)
print(f"f'({x}): {f_prime_x}")

f(x): x^5 + 2x^3 - x^2
f(2): 44
f'(x): 5x^4 + 6x^2 - 2x
f'(2): 100


#### Example 5

In [9]:
polynomial =  {(7, 0)}

print(f"f(x): {to_str(polynomial)}")

x = 2
f_x = evaluate_polynomial(polynomial, x)
print(f"f({x}): {f_x}")

f_prime = differentiate(polynomial)
print(f"f'(x): {to_str(f_prime)}")

f_prime_x = evaluate_polynomial(f_prime, x)
print(f"f'({x}): {f_prime_x}")

f(x): 7
f(2): 7
f'(x): 0
f'(2): 0


 #### Example 6

In [10]:
polynomial =  {(1, 1)}

print(f"f(x): {to_str(polynomial)}")

x = 2
f_x = evaluate_polynomial(polynomial, x)
print(f"f({x}): {f_x}")

f_prime = differentiate(polynomial)
print(f"f'(x): {to_str(f_prime)}")

f_prime_x = evaluate_polynomial(f_prime, x)
print(f"f'({x}): {f_prime_x}")

f(x): x
f(2): 2
f'(x): 1
f'(2): 1


## Reflection

Write your reflection in below cell. With reference to the Rubric for E-tivity 1:
- Provide an accurate description of your code with advantages and disadvantages of design choices.
- Compare your approach to alternative (peer) approaches.
- Clearly describe how you have used your peers' work/input and how this has affected your own understanding / insights.

If you have not used peer input, you may state this, but your submission history in Gitlab should clearly show this is the case.

### Description

#### Data Representations
First I will explain my chosen data representation. 

* I could have used a list where the index in the list prepresents the power however this becomes very inefficient (in terms of memeory and time) as you start to process higher order polynomials e.g. 

    $$2^{1000}$$

    In this case, where you have higher order polynomials, you would have a list with 1000 elements, 999 of which are zero). 

* I have chosen to represent the polynomial as a set of tuples, where each tuple represents an element of the polynomial set. The first element of the tuple is the coefficient and the second element is the power of the corresponding term.  

* Using tuples has many benefits including space efficiency and ease-of-use (i.e. I don't have to write out lots of zeros which may be error-prone).

So, in my case I can represent: 

$$5x^3 + 7x^2 + 3x + 6$$

as:

$${\{(5, 3), (7, 2), (3, 1), (6, 0)\}}$$

#### Implementation
##### Differentition
I implemented the Power rule method for differentiating a polynomial, where the input to the function is a set of tuples (see above) and the output is a set of tuples representing the differentiated polynomial. Currently, it uses a more loop-based approach for ease-of-reading.

##### Evaluating a polynomial
TODO:

#### Advantages/Disadvantages
* using tuples, while efficient does require more knowledge of python data structures. 
* using lists can lead to real inefficiencies as the order and sparseness of the data increases.


#### Peer Assessment

* The list is a nice idea. Simple to understand and very effective for low-rder polynomials. Great for starting and learning how to map mathematics to code.
* Use of recursion is also very elegent. It does have issues as the order or the polynomials increases. For example, you would never use such on low-power moobile devices for crypographic curve evaluation. 
* Mitchell reminded me about the importance of testing more corner cases.
